### PR TITLE
Archive/Restore logos based on subscription privileges

### PR DESCRIPTION
**Pull request by biyeun**
addresses future issues related to: http://manage.dimagi.com/default.asp?189744

on loss of `COMMCARE_LOGO_UPLOADER` privilege on subscription change, remove logos from `logo_refs` and store in `archived_media`.

on gaining the `COMMCARE_LOGO_UPLOADER` privilege, restore any logos that do exist in `archived_media`

### CODE REVIEW

**Comment by proteusvacuum** [comment]
weird that travis didn't run... also, would be good to add some tests for this.

**Comment by biyeun** [comment]
@proteusvacuum added tests

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/accounting/tests/test_subscription_permissions_changes.py, line 1 - F401 'datetime' imported but unused
* corehq/apps/accounting/tests/test_subscription_permissions_changes.py, line 8 - F401 'APP_V2' imported but unused
* corehq/apps/accounting/tests/test_subscription_permissions_changes.py, line 9 - F401 'Module' imported but unused
* corehq/apps/accounting/tests/test_subscription_permissions_changes.py, line 14 - E303 too many blank lines (3)
* corehq/apps/hqmedia/models.py, line 491 - E501 line too long (151 > 115 characters)
* corehq/apps/hqmedia/models.py, line 788 - W391 blank line at end of file

File: corehq/apps/accounting/tests/test_subscription_permissions_changes.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+from corehq.apps.accounting import generator
+from corehq.apps.accounting.models import BillingAccount, DefaultProductPlan, \
+    SoftwarePlanEdition, Subscription
+from corehq.apps.accounting.tests import BaseAccountingTest
+from corehq.apps.app_manager.models import Application
+from corehq.apps.domain.models import Domain
+
+
+
+class TestSubscriptionPermissionsChanges(BaseAccountingTest):
```
Comment (dimagimon) [inline corehq/apps/accounting/tests/test_subscription_permissions_changes.py]: E303 too many blank lines (3)

File: corehq/apps/hqmedia/models.py [hunk 2]
```diff
@@ -487,6 +488,8 @@ class HQMediaMixin(Document):
     # paths to custom logos
     logo_refs = DictProperty()
 
+    archived_media = DictProperty()  # where we store references to the old logos (or other multimedia) on a downgrade, so that information is not lost
```
Comment (dimagimon) [inline corehq/apps/hqmedia/models.py]: E501 line too long (151 > 115 characters)

File: corehq/apps/hqmedia/models.py [hunk 3]
```diff
@@ -756,3 +759,30 @@ def check_media_state(self):
...
+        """
+        has_restored = False
+        if hasattr(self, 'archived_media') and LOGO_ARCHIVE_KEY in self.archived_media:
+            for slug, logo_data in self.archived_media[LOGO_ARCHIVE_KEY].items():
+                self.logo_refs[slug] = logo_data
+                has_restored = True
+                del self.archived_media[LOGO_ARCHIVE_KEY][slug]
+        return has_restored
+
+
```
Comment (dimagimon) [inline corehq/apps/hqmedia/models.py]: W391 blank line at end of file

**Comment by proteusvacuum** [comment]
:+1:

**Comment by biyeun** [comment]
@sravfeyn or interrupt @NoahCarnahan @nickpell

**Pull request merged**